File: doc-examples/example-java/src/main/java/example/PersonFilter.java
package example;

import io.micronaut.serde.PropertyFilter;
import io.micronaut.serde.Serializer;
import jakarta.inject.Named;
import jakarta.inject.Singleton;

@Singleton
@Named("person-filter") // <1>
public class PersonFilter implements PropertyFilter {

    @Override
    public boolean shouldInclude(
        Serializer.EncoderContext encoderContext, Serializer<Object> propertySerializer,
        Object bean, String propertyName, Object propertyValue
    ) {
        if (bean instanceof Person) { // <2>
            Person person = (Person) bean;
            if (propertyName.equals("name")) {
                return person.getPreferredName() == null;
            } else if (propertyName.equals("preferredName")) {
                return person.getPreferredName() != null;
            }
        }
        return true;
    }
}

File: doc-examples/example-python/src/main/python/example/PersonFilter.py
from jakarta.inject import Named, Singleton
from micronaut.serde import PropertyFilter, Serializer
from micronaut.serde.Serializer import EncoderContext

from example.Person import Person


@Singleton
@Named("person-filter")  # <1>
class PersonFilter(PropertyFilter):
    def shouldInclude(
        self,
        encoder_context: EncoderContext,
        property_serializer: Serializer,
        bean: object,
        property_name: str,
        property_value: object,
    ) -> bool:
        if isinstance(bean, Person):  # <2>
            if property_name == "name":
                return bean.preferredName is None
            elif property_name == "preferredName":
                return bean.preferredName is not None
        return True
